crates/holochain/src/conductor/manager/mod.rs

```diff
@@ -97,36 +119,70 @@ pub(crate) async fn keep_alive_task(mut die: broadcast::Receiver<()>) -> Managed
     Ok(())
 }
 
-async fn run(mut new_task_channel: mpsc::Receiver<ManagedTaskAdd>) {
+async fn run(mut new_task_channel: mpsc::Receiver<ManagedTaskAdd>) -> ShutdownResult {
     let mut task_manager = TaskManager::new();
     // Need to have at least one item in the stream or it will exit early
     if let Some(new_task) = new_task_channel.recv().await {
         task_manager.stream.push(new_task);
     } else {
         error!("All senders to task manager were dropped before starting");
-        return;
+        return Err(ShutdownError::TaskManagerFailedToStart);
     }
     loop {
         tokio::select! {
             Some(new_task) = new_task_channel.recv() => {
                 task_manager.stream.push(new_task);
             }
             result = task_manager.stream.next() => match result {
-                Some(Some(new_task)) => task_manager.stream.push(new_task),
-                Some(None) => (),
-                None => break,
+                Some(TaskOutcome::NewTask(new_task)) => task_manager.stream.push(new_task),
+                Some(TaskOutcome::Ignore) => (),
+                Some(TaskOutcome::ExitConductor(error)) => {
+                    let error = match *error {
+                        ManagedTaskError::Join(error) => {
+                            match error.try_into_panic() {
+                                Ok(reason) => {
+                                    // Resume the panic on the main task
+                                    std::panic::resume_unwind(reason);
+                                }
+                                Err(error) => ManagedTaskError::Join(error),
+                            }
+                        }
+                        error => error,
+                    };
+                    error!("Shutting down conductor due to unrecoverable error: {:?}", error);
+                    return Err(ShutdownError::Unrecoverable(error));
+                },
+                None => return Ok(()),
```

Comment:
what's a practical example for this case?

Reply:
The conductor shutting down